File: apps/api/src/models/charts.py
from enum import Enum
from typing import Any

from pydantic import BaseModel


class ChartType(str, Enum):
    BAR = "bar"
    LINE = "line"
    PIE = "pie"


class StyleConfig(BaseModel):
    color: str | None = None
    fill: str | None = None
    opacity: float | None = None
    stroke_width: int | None = None
    stroke_dasharray: str | None = None
    radius: int | None = None


class AxisConfig(BaseModel):
    show: bool = True
    label: str | None = None
    tick_line: bool = True
    tick_margin: int | None = None
    axis_line: bool = True
    grid_lines: bool = True
    formatter: str | None = None  # Could be a string template or function name


class TooltipConfig(BaseModel):
    show: bool = True
    custom_content: str | None = None  # Template or component name
    hide_label: bool = False


class SeriesConfig(BaseModel):
    label: str
    data_key: str
    style: StyleConfig | None = None
    stack_id: str | None = None  # For stacked charts
    type: ChartType | None = None  # For mixed charts


class Chart(BaseModel):
    type: ChartType
    title: str
    description: str | None = None

    # Data configuration
    data: list[dict[str, Any]]

    # Series configuration
    series: list[SeriesConfig]

    # Axes configuration
    x_axis: AxisConfig | None = None
    y_axis: AxisConfig | None = None

    # Visual configuration
    style: StyleConfig | None = None
    tooltip: TooltipConfig | None = None
    legend: bool = True

    # Interaction configuration
    active_index: int | None = None
    animation: bool = True

    # Layout configuration
    width: int | None = None
    height: int | None = None
    margin: dict[str, int] | None = None

    # Additional custom configuration
    config: dict[str, Any] | None = None

File: apps/web/src/app/schemas/charts.ts
import { z } from 'zod';

/** Supported chart types */
export const ChartType = z.enum(['bar', 'line', 'pie', 'donut']);
export type ChartType = z.infer<typeof ChartType>;

/** Styling configuration for chart elements */
export const styleConfigSchema = z
  .object({
    color: z.string().nullable().optional(),
    fill: z.string().nullable().optional(),
    opacity: z.number().nullable().optional().default(0.9),
    stroke_width: z.number().nullable().optional().default(2),
    stroke_dasharray: z.string().nullable().optional(),
    radius: z.number().nullable().optional().default(80),
    inner_radius: z.number().nullable().optional().default(40) // For donut charts
  })
  .nullable()
  .optional();
export type StyleConfig = z.infer<typeof styleConfigSchema>;

/** Configuration for chart tooltips */
export const tooltipConfigSchema = z
  .object({
    show: z.boolean().optional().default(true),
    custom_content: z.string().nullable().optional(),
    hide_label: z.boolean().optional().default(false)
  })
  .nullable()
  .optional();
export type TooltipConfig = z.infer<typeof tooltipConfigSchema>;

/** Configuration for a data series in a chart */
export const seriesConfigSchema = z.object({
  label: z.string(),
  data_key: z.string(),
  style: styleConfigSchema.optional().default({}),
  stack_id: z.string().nullable().optional()
  // Removed type field as it's no longer needed for single-type charts
});
export type SeriesConfig = z.infer<typeof seriesConfigSchema>;

/** Configuration for chart axes */
export const axisConfigSchema = z
  .object({
    show: z.boolean().optional().default(true),
    label: z.string().nullable().optional(),
    tick_line: z.boolean().optional().default(true),
    tick_margin: z.number().optional().default(10),
    axis_line: z.boolean().optional().default(true),
    grid_lines: z.boolean().optional().default(true),
    formatter: z.string().nullable().optional()
  })
  .nullable()
  .optional();
export type AxisConfig = z.infer<typeof axisConfigSchema>;

/** Margin configuration for charts */
export const marginConfigSchema = z
  .object({
    top: z.number().optional().default(0),
    right: z.number().optional().default(0),
    bottom: z.number().optional().default(0),
    left: z.number().optional().default(0)
  })
  .nullable()
  .optional();
export type MarginConfig = z.infer<typeof marginConfigSchema>;

/** Complete chart configuration */
export const chartConfigSchema = z.object({
  type: ChartType,
  title: z.string(),
  description: z.string().nullish(),
  data: z.array(z.record(z.any())),
  series: z.array(seriesConfigSchema),
  x_axis: axisConfigSchema.optional().default({}),
  y_axis: axisConfigSchema.optional().default({}),
  style: styleConfigSchema.optional().default({}),
  tooltip: tooltipConfigSchema.optional().default({}),
  legend: z.boolean().optional().default(true),
  active_index: z.number().nullish(),
  animation: z.boolean().optional().default(true),
  width: z.number().nullish(),
  height: z.number().nullish().default(400),
  margin: marginConfigSchema.optional().default({})
});
export type ChartConfig = z.infer<typeof chartConfigSchema>;
